README
== Description
  A collection of classes that provide different ways to ping computers.

== Prerequisites
  * ffi
  * win32-security (MS Windows only)
  * fakeweb (test only)
  * test-unit (test only)

  Ruby users should use Ruby 1.9.3 or later.
  JRuby users should use JRuby 1.6.7 or later.

== Installation
  gem install net-ping

== Maintainer Wanted!
  I would very much like to turn this gem over to someone who is using it
  more actively than I am. Please let me know if you are interested!

== Notes
  Please read the documentation under the 'doc' directory. Especially pay
  attention to the documentation pertaining to ECONNREFUSED and TCP pings.

  Also note the documentation regarding down hosts.
== How to require net-ping
  You can do either this:

  require 'net/ping'

  In which case you will get Net::Ping and all of its subclasses. Or,
  you can load individual subclasses like this:

  require 'net/ping/tcp'

  The former has the advantage of being easier to remember and all inclusive,
  not to mention backwards compatible. The latter has the advantage of
  reducing your memory footprint.
   
== Known Issues
  Older versions of Ruby 1.9.x may not work with UDP pings.

  Older versions of JRuby will return false positives in UDP pings
  because of an incorrect error class being raised. See JRuby-4896.

  JRuby 1.6.7 or later is required for external pings because of a bug
  in earlier versions with open3 and stream handling.

  ICMP pings will not work with JRuby without some sort of third-party
  library support for raw sockets in Java, such as RockSaw.

== License
  Artistic 2.0

== Contributions
  Although this library is free, please consider having your company
  setup a gittip if used by your company professionally.

  http://www.gittip.com/djberg96/
   
== More documentation
  If you installed this library via Rubygems, you can view the inline
  documentation via ri or fire up 'gem server', and point your browser at
  http://localhost:8808.

== Author
  Daniel J. Berger


CHANGES
== 1.7.6 - 13-Dec-2014
* Changed TCP handling yet again, as it was returning false positives. Thanks
  go to Marcos Piccinini for the spot.
* Now uses relative_require internally.
* Minor tweaks to the gemspec and Rakefile for JRuby.

== 1.7.5 - 2-Nov-2014
* Added packet count and intervals for external pings. Thanks go to Jack Hayter
  for the patch.
* No longer report closed TCP port as active. Thanks go to Joshua Yanovski for
  the spot and patch.

== 1.7.4 - 16-Apr-2014
* Remove the Timeout block for the Ping::HTTP class because it wasn't working
  with JRuby. Instead, we use the builtin open_timeout and read_timeout
  accessors on the underlying http request. Thanks go to Ian Heggie for the
  spot.
* The Ping::HTTP#ping? more robustly parses out the port from the uri argument
  if provided. In addition, the default port is again set back to 80 as a
  last resort in the constructor.
* Added timeout and port tests for the Ping::HTTP class.
* If a host is unreachable explicitly set the result to false regardless of
  the actual exit status. This appears to only affect Windows 7 and later.
* Reinstated the Timeout block for the Ping::ICMP class. Without it, threaded
  pings could end up in an infinite loop. Thanks go to muirmanders for the spot.

== 1.7.3 - 3-Apr-2014
* Removed the Timeout block for the Ping::External class as it apparently
  hasn't worked with open3 for some time. Instead, it now uses your command
  line ping's timeout switch. Thanks go to Andrea Bernardo Ciddio for the
  spot.
* Made ICMP pings thread safe. I think.
* Removed a redundant Timeout block in the Ping::ICMP class.
* If an ICMP ping times out, the exception attribute is set to "timeout".
* Added explicit support for AIX for the Ping::External class.
* Bumped minimum required Ruby version to 1.9.3.

== 1.7.2 - 21-Jan-2014
* Fixed a bug in the Ping::HTTP constructor where it was not honoring the
  port if provided as the second argument. Thanks go to Florian Anderiasch
  for the spot.
* Removed win32-open3 require in the Ping::External class since it isn't
  needed in Ruby 1.9.x or later.
* Refactored the Ping::External class so that it checks the exit status rather
  than try to parse output. Thanks go to Miguel Tubia for the suggestion.
* Dropped explicit support for Windows XP.
* Removed reference to win32-open3 in README.

== 1.7.1 - 26-Sep-2013
* Fixed a bug in the Ping::TCP class where it could return nil instead
  of false. Thanks go to Grandy Nguyen for the patch.

== 1.7.0 - 25-Aug-2013
* Now requires Ruby 1.9.x or later.
* Replaced the main ping code for the Ping::TCP class. The timeout module
  really doesn't work very well with sockets, and so a raw socket approach
  combined with IO.select is now used instead.
* Updates to the doc/ping.txt file.

== 1.6.2 - 8-Aug-2013
* Fixed a bug in the ICMP ping helper module. Thanks go to Yuichi
  Tsunematsu for the spot.
* The win32-security library is not just a development dependency. Thanks
  again go to Yuichi Tsunematsu for the spot.
* Updated the icmp ping tests.
* Updated the gemspec for MS Windows.

== 1.6.1 - 17-Jul-2013
* Automatically set the scheme to "http" if not present when performing
  HTTP pings. See https://bugs.ruby-lang.org/issues/8645 for an issue
  with the uri library that led to this change. Thanks go to Preston Lee
  for the spot.
* Removed the windows/system_info dependency, since it added other
  dependencies just to check the Windows version. Instead, I added a
  helper file that adds some needed Windows functions using FFI.
* Added Rake as a development dependency and updated the Gemfile
  source because Bundler. Thanks to Jean-Philippe Doyle for the patches.
* Added some warning text to the Net::Ping::External tests for users
  who are using OpenDNS. In short, you will probably get test failures
  and surprising true results for bad hosts.

== 1.6.0 - 19-Mar-2013
* Split out the ldap portion of the code into its own branch.
* Don't require resolv-replace on Ruby 1.9.3 or later.
* Now gets proxy information from the http_proxy, https_proxy, and
  no_proxy environment variables and uses that when making requests.
  Thanks go to Kevin Olbrich for the patch.
 
== 1.5.3 - 29-Feb-2012
* Removed the Windows::Console dependency and replaced it with FFI since there
  were only two functions needed. This had the nice side effect of making it
  work with JRuby, too.
* Minor cleanup on some multi-variable assignment.
* Fixed require statements and one test for 1.9.

== 1.5.2 - 4-Nov-2011
* The Net::Ping::HTTP class now defaults to using HEAD instead of GET
  for its ping test. You can alter this via the get_request accessor.
  Thanks got to Balazs Kutil for the patch.
* Added the ssl_verify_mode accessor to the Net::Ping::HTTP class, which
  defaults to VERIFY_NONE. This fixes possible warnings resulting from
  missing SSL certificates when using https. Thanks go to Balazs Kutil
  for the patch.

== 1.5.1 - 15-Sep-2011
* Use RbConfig on Ruby 1.9 and later. Thanks go to Torsten Schönebaum
  for the patch.

== 1.5.0 - 3-May-2011
* Added the Net::Ping::LDAP class which allows you to perform LDAP pings.
  Thanks go to Paul Gallagher for adding this.
* The duration instance variable is now reset to nil on a failed ping. Again,
  thanks go to Paul Gallagher.

== 1.4.1 - 17-Mar-2011
* Incorporated changes from Lukas Zapletal that alters the icmp tests to use
  your localhost instead of going outside the network.
* The http tests were similarly modified courtesy of fakeweb, which is now a
  development dependency. Again, change from Lukas Zapletal.
* The icmp tests were reworked for Windows Vista and later. This required the
  addition of the win32-security library as a development dependency.
* The Net::Ping::WMI class is no longer skipped on require or in tests since
  JRuby now supports win32ole.

== 1.4.0 - 14-Feb-2011
* Added the redirect_limit accessor for the Net::Ping::HTTP class. This caps

lib/net/ping/http.rb
require File.join(File.dirname(__FILE__), 'ping')
require 'net/http'
require 'net/https'
require 'uri'
require 'open-uri'

# Force non-blocking Socket.getaddrinfo on Unix systems. Do not use on
# Windows because it (ironically) causes blocking problems.
unless File::ALT_SEPARATOR or RUBY_VERSION >= "1.9.3"
  require 'resolv-replace'
end

# The Net module serves as a namespace only.
module Net

  # The Ping::HTTP class encapsulates methods for HTTP pings.
  class Ping::HTTP < Ping

    # By default an http ping will follow a redirect and give you the result
    # of the final URI.  If this value is set to false, then it will not
    # follow a redirect and will return false immediately on a redirect.
    #
    attr_accessor :follow_redirect

    # The maximum number of redirects allowed. The default is 5.
    attr_accessor :redirect_limit

    # The user agent used for the HTTP request. The default is nil.
    attr_accessor :user_agent

    # OpenSSL certificate verification mode. The default is VERIFY_NONE.
    attr_accessor :ssl_verify_mode

    # Use GET request instead HEAD. The default is false.
    attr_accessor :get_request

    # was this ping proxied?
    attr_accessor :proxied

    # For unsuccessful requests that return a server error, it is
    # useful to know the HTTP status code of the response.
    attr_reader :code

    # Creates and returns a new Ping::HTTP object. The default port is the
    # port associated with the URI or 80. The default timeout is 5 seconds.
    #
    def initialize(uri=nil, port=nil, timeout=5)
      @follow_redirect = true
      @redirect_limit  = 5
      @ssl_verify_mode = OpenSSL::SSL::VERIFY_NONE
      @get_request     = false
      @code            = nil

      port ||= URI.parse(uri).port if uri
      port ||= 80

      @port = port

      super(uri, port, timeout)
    end

    # Looks for an HTTP response from the URI passed to the constructor.
    # If the result is a kind of Net::HTTPSuccess then the ping was
    # successful and true is returned.  Otherwise, false is returned
    # and the Ping::HTTP#exception method should contain a string
    # indicating what went wrong.
    #
    # If the HTTP#follow_redirect accessor is set to true (which it is
    # by default) and a redirect occurs during the ping, then the
    # HTTP#warning attribute is set to the redirect message, but the
    # return result is still true. If it's set to false then a redirect
    # response is considered a failed ping.
    #
    # If no file or path is specified in the URI, then '/' is assumed.
    # If no scheme is present in the URI, then 'http' is assumed.
    #
    def ping(host = @host)
      super(host)
      bool = false

      # See https://bugs.ruby-lang.org/issues/8645
      host = "http://#{host}" unless host.include?("http")

      uri = URI.parse(host)

      # A port provided here via the host argument overrides anything
      # provided in constructor.
      #
      port = URI.split(host)[3] || URI.parse(host).port || @port
      port = port.to_i

      start_time = Time.now

      response = do_ping(uri, port)

      if response.is_a?(Net::HTTPSuccess)
        bool = true
      elsif redirect?(response) # Check code, HTTPRedirection does not always work
        if @follow_redirect
          @warning = response.message
          rlimit   = 0

          while redirect?(response)
            if rlimit >= redirect_limit
              @exception = "Redirect limit exceeded"
              break
            end
            redirect = URI.parse(response['location'])
            redirect = uri + redirect if redirect.relative?
            response = do_ping(redirect, port)
            rlimit += 1
          end

          if response.is_a?(Net::HTTPSuccess)
            bool = true
          else
            @warning   = nil
            @exception ||= response.message
          end

        else
          @exception = response.message
        end
      else
        @exception ||= response.message
      end

      # There is no duration if the ping failed
      @duration = Time.now - start_time if bool

      bool
    end

    alias ping? ping
    alias pingecho ping
    alias follow_redirect? follow_redirect
    alias uri host
    alias uri= host=

    private

    def redirect?(response)
      response && response.code.to_i >= 300 && response.code.to_i < 400
    end

    def do_ping(uri, port)
      response = nil
      proxy = uri.find_proxy || URI.parse("")

      begin
        uri_path = uri.path.empty? ? '/' : uri.path

        headers = {}
        headers["User-Agent"] = user_agent if user_agent

        http = Net::HTTP::Proxy(proxy.host, proxy.port, proxy.user, proxy.password).new(uri.host, port)

        http.open_timeout = timeout
        http.read_timeout = timeout

        @proxied = http.proxy?

        if @get_request == true
          request = Net::HTTP::Get.new(uri_path)
        else
          request = Net::HTTP::Head.new(uri_path)
        end

        if uri.scheme == 'https'
          http.use_ssl = true
          http.verify_mode = @ssl_verify_mode
        end

        response = http.start{ |h|
          h.open_timeout = timeout
          h.read_timeout = timeout
          h.request(request)
        }
      rescue Exception => err
        @exception = err.message
      end

      @code = response.code if response

      response
    end
  end
end
